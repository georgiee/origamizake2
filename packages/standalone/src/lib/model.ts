export type Vector = [number, number, number];

export type Vertex = {
  cs: Vector,
  i: number
}

export type Face = {
  vs: Vertex[],
  i: number,
  n?: Vector //normal,
  c?: Vector  //centroid,
  es?: any //edges
  ord?: any;
}

export type Edge = {
  raw?: any,
  v1: Vertex,
  v2: Vertex,
  as: string
}

export type EdgeMap = {[s: string]: Edge;}