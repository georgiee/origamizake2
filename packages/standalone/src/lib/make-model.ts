import * as geo from './geom';
import { Face, Vertex, Edge, Vector, EdgeMap } from './model';

export function makeModel({vertices_coords, faces_vertices, edges_vertices, edges_assignment }) {
  // create our vertices object
  const vertices: Vertex[] = vertices_coords.map((value, index) => ({cs: value, i: index} as Vertex));

  // create our faces object (to be enriched later with data)
  const faces:Face[] = faces_vertices.map((facesVerticesIndices, index) => {
    const faceVertices: Vertex[] = facesVerticesIndices.map(index => vertices[index])
    return {vs: faceVertices, i: index} as Face;
  });

  // create our edges objects
  const edges = createEdgeMap({
    vertices,
    edgesData: edges_vertices,
    assignmentData: edges_assignment
  });

  // enricht faces objects
  faces.forEach(face => {
    face.n = createNormal(face);
    face.c = createCentroid(face);
    face.es = createdFaceEdges(face, edges);
  });
}

export function createdFaceEdges(f: Face, edgeMap: EdgeMap): Edge[] {
  const faceVertices = f.vs;
  let edges: Edge[] = faceVertices.map((vertex, index) => {
    const followingVertex = faceVertices[geo.next(index, faceVertices.length)];
    const [a, b] = vertex.i > followingVertex.i ? [followingVertex, vertex] : [vertex, followingVertex];
    const key = "e" + a.i + "e" + b.i;

    const defaultEdge = {
      v1: a,
      v2: b,
      as: 'U'
    };

    let edge = edgeMap[key] || defaultEdge;
    return edge;
  });

  return edges;
}
function createCentroid(face): Vector {
  const faceVertices = face.vs.map(vertex => vertex.cs);
  return geo.centroid(faceVertices);
}

function createNormal(face): Vector {
  const faceVertices = face.vs.map(vertex => vertex.cs);
  return geo.polygonNormal(faceVertices);
}

export function createEdgeMap(
    {vertices, edgesData, assignmentData}: {vertices: Vertex[], edgesData:any, assignmentData: any }
  ): EdgeMap {
  // TODO: Create edges if none are given
  const edges = (edgesData as Array<any>).reduce<EdgeMap>((edgesList: EdgeMap, [edgeVertex1, edgeVertex2], index) => {
    // TODO: Sort edgeVertex1, edgeVertex2 if given in wrong order
    const key = "e" + edgeVertex1 + "e" + edgeVertex2;
    const asssignment = assignmentData[index] || 'U';

    edgesList[key] = {
      raw: edgesData,
      v1: vertices[edgeVertex1],
      v2: vertices[edgeVertex2],
      as: asssignment
    } as Edge;

    return edgesList;
  }, {});

  return edges;
}