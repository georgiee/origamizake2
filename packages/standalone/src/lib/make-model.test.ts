import { createdFaceEdges, createEdgeMap, parseFoldModel, createFaceOrder } from './make-model';
import { Face, Vertex } from './model';
import { SIMPLE_FOLD } from '../folding-examples/simple-fold';

describe('do it', () => {

  const rawEdges = [
    [0, 1],
    [1, 2],
  ]
  const rawAssignments = [
    'V', 'M'
  ]
  const vertices: Vertex[] = [
    { cs: [1, 0, 0], i: 0 },
    { cs: [0, 1, 1], i: 1 },
    { cs: [0, 0, 1], i: 2 }
  ];

  const face: Face = {
    vs: vertices,
    i: 0
  }

  test('createEdgeMap', () => {
    const map = createEdgeMap({vertices, edgesData: rawEdges, assignmentData: rawAssignments})
    expect(map).toMatchSnapshot();
  });

  test('createdFaceEdges', () => {
    const edgeMap = createEdgeMap({vertices, edgesData: rawEdges, assignmentData: rawAssignments})
    const faceEdges = createdFaceEdges(face, edgeMap);
    expect(faceEdges).toMatchSnapshot();
  })

  test('createdFaceEdges with no edgemap', () => {
    const edgeMap = {};
    const faceEdges = createdFaceEdges(face, edgeMap);
    expect(faceEdges).toMatchSnapshot();
  })

  test('createFaceOrders', () => {
    const parsedModel = parseFoldModel(SIMPLE_FOLD);
    parsedModel.faces.forEach(face => {
      createFaceOrder(face, SIMPLE_FOLD.faceOrders, parsedModel.faces);
    });

    expect(parsedModel.faces).toMatchSnapshot();
  })
})