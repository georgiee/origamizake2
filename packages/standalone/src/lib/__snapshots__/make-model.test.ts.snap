// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`do it createEdgeMap 1`] = `
Object {
  "e0e1": Object {
    "as": "V",
    "raw": Array [
      Array [
        0,
        1,
      ],
      Array [
        1,
        2,
      ],
    ],
    "v1": Object {
      "cs": Array [
        1,
        0,
        0,
      ],
      "i": 0,
    },
    "v2": Object {
      "cs": Array [
        0,
        1,
        1,
      ],
      "i": 1,
    },
  },
  "e1e2": Object {
    "as": "M",
    "raw": Array [
      Array [
        0,
        1,
      ],
      Array [
        1,
        2,
      ],
    ],
    "v1": Object {
      "cs": Array [
        0,
        1,
        1,
      ],
      "i": 1,
    },
    "v2": Object {
      "cs": Array [
        0,
        0,
        1,
      ],
      "i": 2,
    },
  },
}
`;

exports[`do it createdFaceEdges 1`] = `
Array [
  Object {
    "as": "V",
    "raw": Array [
      Array [
        0,
        1,
      ],
      Array [
        1,
        2,
      ],
    ],
    "v1": Object {
      "cs": Array [
        1,
        0,
        0,
      ],
      "i": 0,
    },
    "v2": Object {
      "cs": Array [
        0,
        1,
        1,
      ],
      "i": 1,
    },
  },
  Object {
    "as": "M",
    "raw": Array [
      Array [
        0,
        1,
      ],
      Array [
        1,
        2,
      ],
    ],
    "v1": Object {
      "cs": Array [
        0,
        1,
        1,
      ],
      "i": 1,
    },
    "v2": Object {
      "cs": Array [
        0,
        0,
        1,
      ],
      "i": 2,
    },
  },
  Object {
    "as": "U",
    "v1": Object {
      "cs": Array [
        1,
        0,
        0,
      ],
      "i": 0,
    },
    "v2": Object {
      "cs": Array [
        0,
        0,
        1,
      ],
      "i": 2,
    },
  },
]
`;

exports[`do it createdFaceEdges with no edgemap 1`] = `
Array [
  Object {
    "as": "U",
    "v1": Object {
      "cs": Array [
        1,
        0,
        0,
      ],
      "i": 0,
    },
    "v2": Object {
      "cs": Array [
        0,
        1,
        1,
      ],
      "i": 1,
    },
  },
  Object {
    "as": "U",
    "v1": Object {
      "cs": Array [
        0,
        1,
        1,
      ],
      "i": 1,
    },
    "v2": Object {
      "cs": Array [
        0,
        0,
        1,
      ],
      "i": 2,
    },
  },
  Object {
    "as": "U",
    "v1": Object {
      "cs": Array [
        1,
        0,
        0,
      ],
      "i": 0,
    },
    "v2": Object {
      "cs": Array [
        0,
        0,
        1,
      ],
      "i": 2,
    },
  },
]
`;
